File: packages/b2c-tooling-sdk/src/operations/sites/cartridges.ts
/*
 * Copyright (c) 2025, Salesforce, Inc.
 * SPDX-License-Identifier: Apache-2
 * For full license text, see the license.txt file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
 */
/**
 * Site cartridge path operations for B2C Commerce instances.
 *
 * Provides functions for managing the ordered list of active cartridges
 * on a site via SCAPI with temporary OCAPI fallback, plus site archive
 * import/export when neither direct API is available.
 */
import JSZip from 'jszip';
import type {B2CInstance} from '../../instance/index.js';
import {getLogger} from '../../logging/logger.js';
import {siteArchiveImport, siteArchiveExportToBuffer} from '../jobs/site-archive.js';
import type {WaitForJobOptions} from '../jobs/run.js';
import {createSitesBackend} from './sites-backend.js';
import type {CartridgePosition} from './sites-types.js';

/** The special site ID for Business Manager. */
export const BM_SITE_ID = 'Sites-Site';

/** Position options for adding a cartridge. */
export type {CartridgePosition} from './sites-types.js';

/** Options for adding a cartridge to a site's cartridge path. */
export interface AddCartridgeOptions {
  /** Cartridge name to add. */
  name: string;
  /** Position to add the cartridge (default: 'first'). */
  position: CartridgePosition;
  /** Target cartridge name (required when position is 'before' or 'after'). */
  target?: string;
}

/** Options for cartridge path update operations that may run jobs. */
export interface CartridgeUpdateOptions {
  /** Callback for operation-level status messages (e.g. "Exporting site preferences..."). */
  log?: (message: string) => void;
  /** Wait options for underlying job execution (polling interval, timeout, progress). */
  waitOptions?: WaitForJobOptions;
}

/** Result of a cartridge path operation. */
export interface CartridgePathResult {
  /** Site ID. */
  siteId: string;
  /** Colon-separated cartridge path string. */
  cartridges: string;
  /** Cartridge names as an ordered array. */
  cartridgeList: string[];
}

/**
 * Parses a colon-separated cartridge path string into a CartridgePathResult.
 */
function toResult(siteId: string, cartridges: string): CartridgePathResult {
  const trimmed = cartridges.trim();
  return {
    siteId,
    cartridges: trimmed,
    cartridgeList: trimmed ? trimmed.split(':') : [],
  };
}

/**
 * Gets the cartridge path for a site.
 *
 * Uses the configured Sites backend to read the cartridge path. Auto mode
 * prefers SCAPI and temporarily falls back to OCAPI.
 * Works for all sites including Business Manager (Sites-Site).
 *
 * @param instance - B2C instance to query
 * @param siteId - Site ID (e.g. 'RefArch', 'Sites-Site')
 * @returns Cartridge path result
 *
 * @example
 * ```typescript
 * const result = await getCartridgePath(instance, 'RefArch');
 * console.log(result.cartridgeList); // ['app_storefront_base', 'plugin_applepay']
 *
 * // Business Manager
 * const bmResult = await getCartridgePath(instance, 'Sites-Site');
 * ```
 */
export async function getCartridgePath(instance: B2CInstance, siteId: string): Promise<CartridgePathResult> {
  try {
    const cartridges = await createSitesBackend({instance}).getCartridgePath(siteId);
    return toResult(siteId, cartridges);
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    throw new Error(`Failed to get cartridge path for site "${siteId}": ${message}`, {cause: error});
  }
}

/**
 * Adds a cartridge to a site's cartridge path.
 *
 * For regular sites, uses the SCAPI-first Sites backend and falls back to site
 * archive import if neither direct backend is available.
 * For Business Manager (Sites-Site), always uses site archive import.
 *
 * @param instance - B2C instance
 * @param siteId - Site ID
 * @param options - Cartridge name, position, and optional target
 * @returns Updated cartridge path
 *
 * @example
 * ```typescript
 * // Add to beginning (default)
 * await addCartridge(instance, 'RefArch', { name: 'my_cartridge', position: 'first' });
 *
 * // Add before a specific cartridge
 * await addCartridge(instance, 'RefArch', {
 *   name: 'my_cartridge', position: 'before', target: 'app_storefront_base'
 * });
 *
 * // Business Manager
 * await addCartridge(instance, 'Sites-Site', { name: 'bm_ext', position: 'first' });
 * ```
 */
export async function addCartridge(
  instance: B2CInstance,
  siteId: string,
  options: AddCartridgeOptions,
  updateOptions?: CartridgeUpdateOptions,
): Promise<CartridgePathResult> {
  const logger = getLogger();

  // BM always uses import/export
  if (siteId === BM_SITE_ID) {
    logger.debug({siteId}, 'Business Manager site — using site archive import for cartridge add');
    return addCartridgeViaImport(instance, siteId, options, updateOptions);
  }

  const backend = createSitesBackend({instance});
  try {
    return toResult(siteId, await backend.addCartridge(siteId, options.name, options.position, options.target));
  } catch (backendError) {
    return handleFallback(instance, siteId, 'add', backendError, () =>
      addCartridgeViaImport(instance, siteId, options, updateOptions),
    );
  }
}

/**
 * Removes a cartridge from a site's cartridge path.
 *
 * For regular sites, uses the SCAPI-first Sites backend and falls back to site
 * archive import if neither direct backend is available.
 * For Business Manager (Sites-Site), always uses site archive import.
 *
 * @param instance - B2C instance
 * @param siteId - Site ID
 * @param cartridgeName - Name of the cartridge to remove
 * @returns Updated cartridge path
 *
 * @example
 * ```typescript
 * await removeCartridge(instance, 'RefArch', 'old_cartridge');
 * ```
 */
export async function removeCartridge(
  instance: B2CInstance,
  siteId: string,
  cartridgeName: string,
  updateOptions?: CartridgeUpdateOptions,
): Promise<CartridgePathResult> {
  const logger = getLogger();

  if (siteId === BM_SITE_ID) {
    logger.debug({siteId}, 'Business Manager site — using site archive import for cartridge remove');
    return removeCartridgeViaImport(instance, siteId, cartridgeName, updateOptions);
  }

  const backend = createSitesBackend({instance});
  try {
    return toResult(siteId, await backend.removeCartridge(siteId, cartridgeName));
  } catch (backendError) {
    return handleFallback(instance, siteId, 'remove', backendError, () =>
      removeCartridgeViaImport(instance, siteId, cartridgeName, updateOptions),
    );
  }
}

/**
 * Replaces the entire cartridge path for a site.
 *
 * For regular sites, uses the SCAPI-first Sites backend and falls back to site
 * archive import if neither direct backend is available.
 * For Business Manager (Sites-Site), always uses site archive import.
 *
 * @param instance - B2C instance
 * @param siteId - Site ID
 * @param cartridges - New cartridge path (colon-separated string)
 * @returns Updated cartridge path
 *
 * @example
 * ```typescript
 * await setCartridgePath(instance, 'RefArch', 'app_storefront_base:plugin_applepay');
 * ```
 */
export async function setCartridgePath(
  instance: B2CInstance,
  siteId: string,
  cartridges: string,
  updateOptions?: CartridgeUpdateOptions,
): Promise<CartridgePathResult> {
  const logger = getLogger();

  if (siteId === BM_SITE_ID) {
    logger.debug({siteId}, 'Business Manager site — using site archive import for cartridge set');
    return setCartridgePathViaImport(instance, siteId, cartridges, updateOptions);
  }

  const backend = createSitesBackend({instance});
  try {
    return toResult(siteId, await backend.setCartridgePath(siteId, cartridges));
  } catch (backendError) {
    return handleFallback(instance, siteId, 'set', backendError, () =>
      setCartridgePathViaImport(instance, siteId, cartridges, updateOptions),
    );
  }
}

// ---------------------------------------------------------------------------
// Internal: Fallback handler
// ---------------------------------------------------------------------------

async function handleFallback(
  instance: B2CInstance,
  siteId: string,
  operation: string,
  backendError: unknown,
  fallbackFn: () => Promise<CartridgePathResult>,
): Promise<CartridgePathResult> {
  const logger = getLogger();
  const backendMessage = backendError instanceof Error ? backendError.message : String(backendError);

  logger.warn(
    {siteId, operation, error: backendMessage},
    `Direct API ${operation} failed, trying site archive import fallback`,
  );

  try {
    return await fallbackFn();
  } catch (importError) {
    const importMessage = importError instanceof Error ? importError.message : String(importError);
    throw new Error(
      [
        `Failed to ${operation} cartridge path for site "${siteId}".`,
        '',
        `SCAPI/OCAPI direct update failed: ${backendMessage}`,
        `Site archive import fallback also failed: ${importMessage}`,
        '',
        'To fix, configure one of:',
        '  • SCAPI Sites API: Grant sfcc.sites.rw',
        '  • OCAPI Data API: Grant POST/PUT/DELETE on /sites/*/cartridges',
        '  • Site import: Grant job execution permissions for sfcc-site-archive-import and WebDAV write access to Impex/',
        '',
        'See: https://salesforcecommercecloud.github.io/b2c-developer-tooling/guide/authentication.html',
      ].join('\n'),
      {cause: importError},
    );
  }
}

// ---------------------------------------------------------------------------
// Internal: Import/export-based operations
// ---------------------------------------------------------------------------

/**
 * Reads the current cartridge path via site archive export.
 * Used as a fallback when OCAPI GET also fails.
 */
async function getCartridgePathViaExport(
  instance: B2CInstance,
  siteId: string,
  updateOptions?: CartridgeUpdateOptions,
): Promise<string> {
  const logger = getLogger();
  const {log, waitOptions} = updateOptions ?? {};
  logger.debug({siteId}, 'Reading cartridge path via site archive export');
  log?.(
    `Exporting ${siteId === BM_SITE_ID ? 'organization preferences' : 'site descriptor'} to read cartridge path...`,
  );

  if (siteId === BM_SITE_ID) {
    const result = await siteArchiveExportToBuffer(instance, {global_data: {preferences: true}}, {waitOptions});
    const zip = await JSZip.loadAsync(result.data);
    const prefsXml = await findFileInZip(zip, 'preferences.xml');
    if (!prefsXml) {
      throw new Error('preferences.xml not found in export archive');
    }
    return parseBmCartridgesFromPreferencesXml(prefsXml);
  }

  const result = await siteArchiveExportToBuffer(instance, {sites: {[siteId]: {site_descriptor: true}}}, {waitOptions});
  const zip = await JSZip.loadAsync(result.data);
  const descriptorXml = await findFileInZip(zip, 'site.xml');
  if (!descriptorXml) {
    throw new Error(`site.xml not found in export archive for site "${siteId}"`);
  }
  return parseSiteCartridgesFromDescriptorXml(descriptorXml);
}

async function setCartridgePathViaImport(
  instance: B2CInstance,
  siteId: string,
  cartridges: string,
  updateOptions?: CartridgeUpdateOptions,
): Promise<CartridgePathResult> {
  const logger = getLogger();
  const {log, waitOptions} = updateOptions ?? {};
  logger.debug({siteId, cartridges}, 'Setting cartridge path via site archive import');
  log?.('Importing updated cartridge path...');

  const zip = new JSZip();

  if (siteId === BM_SITE_ID) {
    zip.file('preferences.xml', generateBmPreferencesXml(cartridges));
  } else {
    const sitesFolder = zip.folder(`sites/${siteId}`)!;
    sitesFolder.file('site.xml', generateSiteDescriptorXml(siteId, cartridges));
  }

  const buffer = await zip.generateAsync({type: 'nodebuffer', compression: 'DEFLATE'});
  await siteArchiveImport(instance, buffer, {waitOptions});

  return toResult(siteId, cartridges);
}

async function addCartridgeViaImport(
  instance: B2CInstance,
  siteId: string,
  options: AddCartridgeOptions,
  updateOptions?: CartridgeUpdateOptions,
): Promise<CartridgePathResult> {
  // Read current path
  let currentPath: string;
  try {
    const result = await getCartridgePath(instance, siteId);
    currentPath = result.cartridges;
  } catch {
    currentPath = await getCartridgePathViaExport(instance, siteId, updateOptions);
  }

  const cartridgeList = currentPath ? currentPath.split(':') : [];

  // Check if already exists
  if (cartridgeList.includes(options.name)) {
    throw new Error(`Cartridge "${options.name}" already exists in the cartridge path for site "${siteId}"`);
  }

  // Apply position logic
  const newList = applyCartridgePosition(cartridgeList, options);

  return setCartridgePathViaImport(instance, siteId, newList.join(':'), updateOptions);
}

async function removeCartridgeViaImport(
  instance: B2CInstance,
  siteId: string,
  cartridgeName: string,
  updateOptions?: CartridgeUpdateOptions,
): Promise<CartridgePathResult> {
  let currentPath: string;
  try {
    const result = await getCartridgePath(instance, siteId);
    currentPath = result.cartridges;
  } catch {
    currentPath = await getCartridgePathViaExport(instance, siteId, updateOptions);
  }

  const cartridgeList = currentPath ? currentPath.split(':') : [];
  const index = cartridgeList.indexOf(cartridgeName);
  if (index === -1) {
    throw new Error(`Cartridge "${cartridgeName}" not found in the cartridge path for site "${siteId}"`);
  }

  cartridgeList.splice(index, 1);
  return setCartridgePathViaImport(instance, siteId, cartridgeList.join(':'), updateOptions);
}

// ---------------------------------------------------------------------------
// Internal: Cartridge position logic
// ---------------------------------------------------------------------------

function applyCartridgePosition(cartridgeList: string[], options: AddCartridgeOptions): string[] {
  const list = [...cartridgeList];
  const {name, position, target} = options;

  switch (position) {
    case 'first':
      list.unshift(name);
      break;
    case 'last':
      list.push(name);
      break;
    case 'before': {
      const idx = list.indexOf(target!);
      if (idx === -1) {
        throw new Error(`Target cartridge "${target}" not found in the cartridge path`);
      }
      list.splice(idx, 0, name);
      break;
    }
    case 'after': {
      const idx = list.indexOf(target!);
      if (idx === -1) {
        throw new Error(`Target cartridge "${target}" not found in the cartridge path`);
      }
      list.splice(idx + 1, 0, name);
      break;
    }
  }

  return list;
}

// ---------------------------------------------------------------------------
// Internal: XML generation and parsing
// ---------------------------------------------------------------------------

function generateBmPreferencesXml(cartridges: string): string {
  return `<?xml version="1.0" encoding="UTF-8"?>
<preferences xmlns="http://www.demandware.com/xml/impex/preferences/2007-03-31">
    <standard-preferences>
        <all-instances>
            <preference preference-id="CustomCartridges">${escapeXml(cartridges)}</preference>
        </all-instances>
    </standard-preferences>
</preferences>
`;
}

function generateSiteDescriptorXml(siteId: string, cartridges: string): string {
  return `<?xml version="1.0" encoding="UTF-8"?>
<site xmlns="http://www.demandware.com/xml/impex/site/2006-10-31" site-id="${escapeXml(siteId)}">
    <cartridges>${escapeXml(cartridges)}</cartridges>
</site>
`;
}

function parseBmCartridgesFromPreferencesXml(xml: string): string {
  // Extract CustomCartridges preference value
  const match = xml.match(/<preference\s+preference-id="CustomCartridges"[^>]*>([^<]*)<\/preference>/);
  return match?.[1]?.trim() ?? '';
}

function parseSiteCartridgesFromDescriptorXml(xml: string): string {
  // Extract cartridges element value
  const match = xml.match(/<cartridges>([^<]*)<\/cartridges>/);
  return match?.[1]?.trim() ?? '';
}

function escapeXml(str: string): string {
  return str.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;');
}

async function findFileInZip(zip: JSZip, filename: string): Promise<string | null> {
  for (const [path, entry] of Object.entries(zip.files)) {
    if (!entry.dir && path.endsWith(`/${filename}`)) {
      return entry.async('text');
    }
  }
  return null;
}

File: python/b2c-tooling-sdk/src/b2c_tooling_sdk/operations/sites/cartridges.py
# Copyright (c) 2025, Salesforce, Inc.
# SPDX-License-Identifier: Apache-2.0
# For full license text, see the LICENSE file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
"""Site cartridge path operations for B2C Commerce instances.

Mirrors ``src/operations/sites/cartridges.ts``. Provides functions for
managing the ordered list of active cartridges on a site via SCAPI with
temporary OCAPI fallback, plus site archive import/export when neither direct
API is available.

Cross-module note: this module needs ``site_archive_import`` and
``site_archive_export_to_buffer`` from ``operations.jobs.site_archive`` for its
fallback path. To avoid a hard import-time dependency on ``operations.jobs``
(which imports back into ``operations.sites``), the import is deferred to
:func:`_import_jobs_site_archive`, called only when the fallback path actually
runs. Tests monkeypatch that single function instead of the real jobs module.
"""

from __future__ import annotations

import io
import re
import zipfile
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

from b2c_tooling_sdk.logging import get_logger
from b2c_tooling_sdk.operations.sites.sites_backend import SitesBackendConfig, create_sites_backend
from b2c_tooling_sdk.operations.sites.sites_types import CartridgePosition

if TYPE_CHECKING:
    from b2c_tooling_sdk.instance import B2CInstance
    from b2c_tooling_sdk.operations.jobs.run import WaitForJobOptions

#: The special site ID for Business Manager.
BM_SITE_ID = "Sites-Site"


@dataclass
class AddCartridgeOptions:
    """Options for adding a cartridge to a site's cartridge path."""

    #: Cartridge name to add.
    name: str
    #: Position to add the cartridge (default: ``"first"``).
    position: CartridgePosition = "first"
    #: Target cartridge name (required when position is ``"before"`` or ``"after"``).
    target: str | None = None


@dataclass
class CartridgeUpdateOptions:
    """Options for cartridge path update operations that may run jobs."""

    #: Callback for operation-level status messages (e.g. "Exporting site preferences...").
    log: Callable[[str], None] | None = None
    #: Wait options for underlying job execution (polling interval, timeout, progress).
    wait_options: WaitForJobOptions | None = None


@dataclass
class CartridgePathResult:
    """Result of a cartridge path operation."""

    #: Site ID.
    site_id: str
    #: Colon-separated cartridge path string.
    cartridges: str
    #: Cartridge names as an ordered array.
    cartridge_list: list[str]


def _import_jobs_site_archive() -> tuple[
    Callable[..., Awaitable[Any]],
    Callable[..., Awaitable[Any]],
]:
    """Deferred import indirection for ``operations.jobs.site_archive``.

    Kept as a standalone module-level function (rather than an inline import at
    each call site) so tests can monkeypatch it directly — ``operations.jobs``
    never needs to exist for cartridge-path tests that don't exercise the
    fallback path, and tests that do exercise it never need the real jobs
    module to be importable.

    :returns: ``(site_archive_import, site_archive_export_to_buffer)``.
    """
    from b2c_tooling_sdk.operations.jobs.site_archive import (  # noqa: PLC0415
        site_archive_export_to_buffer,
        site_archive_import,
    )

    return site_archive_import, site_archive_export_to_buffer


def _to_result(site_id: str, cartridges: str) -> CartridgePathResult:
    """Parses a colon-separated cartridge path string into a :class:`CartridgePathResult`."""
    trimmed = cartridges.strip()
    return CartridgePathResult(
        site_id=site_id,
        cartridges=trimmed,
        cartridge_list=trimmed.split(":") if trimmed else [],
    )


async def get_cartridge_path(instance: B2CInstance, site_id: str) -> CartridgePathResult:
    """Gets the cartridge path for a site.

    Uses the configured Sites backend to read the cartridge path. Auto mode
    prefers SCAPI and temporarily falls back to OCAPI. Works for all sites
    including Business Manager (``Sites-Site``).

    :param instance: B2C instance to query.
    :param site_id: Site ID (e.g. ``"RefArch"``, ``"Sites-Site"``).
    :returns: Cartridge path result.
    """
    try:
        cartridges = await create_sites_backend(SitesBackendConfig(instance=instance)).get_cartridge_path(site_id)
        return _to_result(site_id, cartridges)
    except Exception as error:
        raise RuntimeError(f'Failed to get cartridge path for site "{site_id}": {error}') from error


async def add_cartridge(
    instance: B2CInstance,
    site_id: str,
    options: AddCartridgeOptions,
    update_options: CartridgeUpdateOptions | None = None,
) -> CartridgePathResult:
    """Adds a cartridge to a site's cartridge path.

    For regular sites, uses the SCAPI-first Sites backend and falls back to
    site archive import if neither direct backend is available. For Business
    Manager (``Sites-Site``), always uses site archive import.
    """
    logger = get_logger()

    # BM always uses import/export
    if site_id == BM_SITE_ID:
        logger.debug('Business Manager site "%s" — using site archive import for cartridge add', site_id)
        return await _add_cartridge_via_import(instance, site_id, options, update_options)

    backend = create_sites_backend(SitesBackendConfig(instance=instance))
    try:
        cartridges = await backend.add_cartridge(site_id, options.name, options.position, options.target)
        return _to_result(site_id, cartridges)
    except Exception as backend_error:
        return await _handle_fallback(
            site_id,
            "add",
            backend_error,
            lambda: _add_cartridge_via_import(instance, site_id, options, update_options),
        )


async def remove_cartridge(
    instance: B2CInstance,
    site_id: str,
    cartridge_name: str,
    update_options: CartridgeUpdateOptions | None = None,
) -> CartridgePathResult:
    """Removes a cartridge from a site's cartridge path.

    For regular sites, uses the SCAPI-first Sites backend and falls back to
    site archive import if neither direct backend is available. For Business
    Manager (``Sites-Site``), always uses site archive import.
    """
    logger = get_logger()

    if site_id == BM_SITE_ID:
        logger.debug('Business Manager site "%s" — using site archive import for cartridge remove', site_id)
        return await _remove_cartridge_via_import(instance, site_id, cartridge_name, update_options)

    backend = create_sites_backend(SitesBackendConfig(instance=instance))
    try:
        cartridges = await backend.remove_cartridge(site_id, cartridge_name)
        return _to_result(site_id, cartridges)
    except Exception as backend_error:
        return await _handle_fallback(
            site_id,
            "remove",
            backend_error,
            lambda: _remove_cartridge_via_import(instance, site_id, cartridge_name, update_options),
        )


async def set_cartridge_path(
    instance: B2CInstance,
    site_id: str,
    cartridges: str,
    update_options: CartridgeUpdateOptions | None = None,
) -> CartridgePathResult:
    """Replaces the entire cartridge path for a site.

    For regular sites, uses the SCAPI-first Sites backend and falls back to
    site archive import if neither direct backend is available. For Business
    Manager (``Sites-Site``), always uses site archive import.
    """
    logger = get_logger()

    if site_id == BM_SITE_ID:
        logger.debug('Business Manager site "%s" — using site archive import for cartridge set', site_id)
        return await _set_cartridge_path_via_import(instance, site_id, cartridges, update_options)

    backend = create_sites_backend(SitesBackendConfig(instance=instance))
    try:
        result_cartridges = await backend.set_cartridge_path(site_id, cartridges)
        return _to_result(site_id, result_cartridges)
    except Exception as backend_error:
        return await _handle_fallback(
            site_id,
            "set",
            backend_error,
            lambda: _set_cartridge_path_via_import(instance, site_id, cartridges, update_options),
        )


# ---------------------------------------------------------------------------
# Internal: Fallback handler
# ---------------------------------------------------------------------------


async def _handle_fallback(
    site_id: str,
    operation: str,
    backend_error: BaseException,
    fallback_fn: Callable[[], Awaitable[CartridgePathResult]],
) -> CartridgePathResult:
    logger = get_logger()
    backend_message = str(backend_error)

    logger.warning(
        'Direct API %s failed for site "%s" (%s), trying site archive import fallback',
        operation,
        site_id,
        backend_message,
    )

    try:
        return await fallback_fn()
    except Exception as import_error:
        import_message = str(import_error)
        message = "\n".join(
            [
                f'Failed to {operation} cartridge path for site "{site_id}".',
                "",
                f"SCAPI/OCAPI direct update failed: {backend_message}",
                f"Site archive import fallback also failed: {import_message}",
                "",
                "To fix, configure one of:",
                "  - SCAPI Sites API: Grant sfcc.sites.rw",
                "  - OCAPI Data API: Grant POST/PUT/DELETE on /sites/*/cartridges",
                "  - Site import: Grant job execution permissions for sfcc-site-archive-import "
                "and WebDAV write access to Impex/",
                "",
                "See: https://salesforcecommercecloud.github.io/b2c-developer-tooling/guide/authentication.html",
            ]
        )
        raise RuntimeError(message) from import_error


# ---------------------------------------------------------------------------
# Internal: Import/export-based operations
# ---------------------------------------------------------------------------


async def _get_cartridge_path_via_export(
    instance: B2CInstance,
    site_id: str,
    update_options: CartridgeUpdateOptions | None = None,
) -> str:
    """Reads the current cartridge path via site archive export.

    Used as a fallback when OCAPI GET also fails.
    """
    logger = get_logger()
    opts = update_options or CartridgeUpdateOptions()
    logger.debug('Reading cartridge path via site archive export for site "%s"', site_id)
    if opts.log:
        target_desc = "organization preferences" if site_id == BM_SITE_ID else "site descriptor"
        opts.log(f"Exporting {target_desc} to read cartridge path...")

    _, site_archive_export_to_buffer = _import_jobs_site_archive()

    if site_id == BM_SITE_ID:
        result = await site_archive_export_to_buffer(
            instance, {"global_data": {"preferences": True}}, wait_options=opts.wait_options
        )
        prefs_xml = _find_file_in_zip(result.data, "preferences.xml")
        if prefs_xml is None:
            raise RuntimeError("preferences.xml not found in export archive")
        return _parse_bm_cartridges_from_preferences_xml(prefs_xml)

    result = await site_archive_export_to_buffer(
        instance, {"sites": {site_id: {"site_descriptor": True}}}, wait_options=opts.wait_options
    )
    descriptor_xml = _find_file_in_zip(result.data, "site.xml")
    if descriptor_xml is None:
        raise RuntimeError(f'site.xml not found in export archive for site "{site_id}"')
    return _parse_site_cartridges_from_descriptor_xml(descriptor_xml)


async def _set_cartridge_path_via_import(
    instance: B2CInstance,
    site_id: str,
    cartridges: str,
    update_options: CartridgeUpdateOptions | None = None,
) -> CartridgePathResult:
    logger = get_logger()
    opts = update_options or CartridgeUpdateOptions()
    logger.debug('Setting cartridge path via site archive import for site "%s": %s', site_id, cartridges)
    if opts.log:
        opts.log("Importing updated cartridge path...")

    if site_id == BM_SITE_ID:
        buffer = _build_zip({"preferences.xml": _generate_bm_preferences_xml(cartridges)})
    else:
        buffer = _build_zip({f"sites/{site_id}/site.xml": _generate_site_descriptor_xml(site_id, cartridges)})

    site_archive_import, _ = _import_jobs_site_archive()
    await site_archive_import(instance, buffer, wait_options=opts.wait_options)

    return _to_result(site_id, cartridges)


async def _add_cartridge_via_import(
    instance: B2CInstance,
    site_id: str,
    options: AddCartridgeOptions,
    update_options: CartridgeUpdateOptions | None = None,
) -> CartridgePathResult:
    # Read current path
    try:
        current_path = (await get_cartridge_path(instance, site_id)).cartridges
    except Exception:
        current_path = await _get_cartridge_path_via_export(instance, site_id, update_options)

    cartridge_list = current_path.split(":") if current_path else []

    # Check if already exists
    if options.name in cartridge_list:
        raise RuntimeError(f'Cartridge "{options.name}" already exists in the cartridge path for site "{site_id}"')

    # Apply position logic
    new_list = _apply_cartridge_position(cartridge_list, options)

    return await _set_cartridge_path_via_import(instance, site_id, ":".join(new_list), update_options)


async def _remove_cartridge_via_import(
    instance: B2CInstance,
    site_id: str,
    cartridge_name: str,
    update_options: CartridgeUpdateOptions | None = None,
) -> CartridgePathResult:
    try:
        current_path = (await get_cartridge_path(instance, site_id)).cartridges
    except Exception:
        current_path = await _get_cartridge_path_via_export(instance, site_id, update_options)

    cartridge_list = current_path.split(":") if current_path else []
    if cartridge_name not in cartridge_list:
        raise RuntimeError(f'Cartridge "{cartridge_name}" not found in the cartridge path for site "{site_id}"')

    cartridge_list.remove(cartridge_name)
    return await _set_cartridge_path_via_import(instance, site_id, ":".join(cartridge_list), update_options)


# ---------------------------------------------------------------------------
# Internal: Cartridge position logic
# ---------------------------------------------------------------------------


def _apply_cartridge_position(cartridge_list: list[str], options: AddCartridgeOptions) -> list[str]:
    result = list(cartridge_list)
    name, position, target = options.name, options.position, options.target

    if position == "first":
        result.insert(0, name)
    elif position == "last":
        result.append(name)
    elif position == "before":
        if target is None or target not in result:
            raise RuntimeError(f'Target cartridge "{target}" not found in the cartridge path')
        result.insert(result.index(target), name)
    elif position == "after":
        if target is None or target not in result:
            raise RuntimeError(f'Target cartridge "{target}" not found in the cartridge path')
        result.insert(result.index(target) + 1, name)

    return result


# ---------------------------------------------------------------------------
# Internal: XML generation and parsing
# ---------------------------------------------------------------------------

_PREFERENCE_RE = re.compile(r'<preference\s+preference-id="CustomCartridges"[^>]*>([^<]*)</preference>')
_CARTRIDGES_RE = re.compile(r"<cartridges>([^<]*)</cartridges>")


def _generate_bm_preferences_xml(cartridges: str) -> str:
    escaped = _escape_xml(cartridges)
    return (
        '<?xml version="1.0" encoding="UTF-8"?>\n'
        '<preferences xmlns="http://www.demandware.com/xml/impex/preferences/2007-03-31">\n'
        "    <standard-preferences>\n"
        "        <all-instances>\n"
        f'            <preference preference-id="CustomCartridges">{escaped}</preference>\n'
        "        </all-instances>\n"
        "    </standard-preferences>\n"
        "</preferences>\n"
    )


def _generate_site_descriptor_xml(site_id: str, cartridges: str) -> str:
    escaped_site_id = _escape_xml(site_id)
    escaped_cartridges = _escape_xml(cartridges)
    return (
        '<?xml version="1.0" encoding="UTF-8"?>\n'
        f'<site xmlns="http://www.demandware.com/xml/impex/site/2006-10-31" site-id="{escaped_site_id}">\n'
        f"    <cartridges>{escaped_cartridges}</cartridges>\n"
        "</site>\n"
    )


def _parse_bm_cartridges_from_preferences_xml(xml: str) -> str:
    """Extracts the ``CustomCartridges`` preference value."""
    match = _PREFERENCE_RE.search(xml)
    return match.group(1).strip() if match else ""


def _parse_site_cartridges_from_descriptor_xml(xml: str) -> str:
    """Extracts the ``cartridges`` element value."""
    match = _CARTRIDGES_RE.search(xml)
    return match.group(1).strip() if match else ""


def _escape_xml(value: str) -> str:
    return value.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;").replace('"', "&quot;")


def _build_zip(files: dict[str, str]) -> bytes:
    """Builds an in-memory zip archive (JSZip equivalent) with the given ``path -> text`` entries."""
    buffer = io.BytesIO()
    with zipfile.ZipFile(buffer, "w", zipfile.ZIP_DEFLATED) as zip_file:
        for path, content in files.items():
            zip_file.writestr(path, content)
    return buffer.getvalue()


def _find_file_in_zip(zip_bytes: bytes, filename: str) -> str | None:
    """Finds the first non-directory entry whose path ends with ``/{filename}``."""
    with zipfile.ZipFile(io.BytesIO(zip_bytes)) as zip_file:
        for name in zip_file.namelist():
            if not name.endswith("/") and name.endswith(f"/{filename}"):
                return zip_file.read(name).decode("utf-8")
    return None


__all__ = [
    "BM_SITE_ID",
    "AddCartridgeOptions",
    "CartridgePathResult",
    "CartridgeUpdateOptions",
    "add_cartridge",
    "get_cartridge_path",
    "remove_cartridge",
    "set_cartridge_path",
]
